core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationLifespanTest.java

```diff
@@ -289,4 +303,37 @@ public void testPrimaryNotExpiredButBackupWas() throws InterruptedException, Exe
       assertEquals(value + "-expire-backup", cache1.get(key));
       assertEquals(value + "-expire-backup", cache2.get(key));
    }
+
+   public void testExpirationWithNoValueEquals() {
+      Object key = createKey(cache0, cache1);
+      cache0.put(key, new ExpirationFunctionalTest.NoEquals("value"), 10, TimeUnit.MINUTES);
+
+      assertEquals(1, cache0.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+      assertEquals(1, cache1.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+
+      // Now we expire on cache0, it should still exist on cache1
+      ts0.advance(TimeUnit.MINUTES.toMillis(10) + 1);
+      ts1.advance(TimeUnit.MINUTES.toMillis(10) + 1);
+
+      assertEquals(1, cache0.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+      assertEquals(1, cache1.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+
+      cache0.getAdvancedCache().getExpirationManager().processExpiration();
+
+      // Both should be empty
+      if (cacheMode != CacheMode.SCATTERED_SYNC) {
+         assertEquals(0, cache0.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+         assertEquals(0, cache1.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+      } else {
+         verifyNoValue(cache0.getAdvancedCache().getDataContainer().iteratorIncludingExpired());
+         verifyNoValue(cache1.getAdvancedCache().getDataContainer().iteratorIncludingExpired());
```

Comment:
Is this because the scattered caches may keep entries with RemoteMetadata pointing to other nodes even when the entry is expired on those nodes? I thought this can happen only after a membership change.

Reply:
TBH I didn't investigate it much. For whatever reason an entry is left in the DataContainer with a null value. It seems like a bug, but given the fact that we are dropping Scattered cache I didn't invest any time to looking closer.

Reply:
Fair enough!